### PR TITLE
Replace several activation script snippets with declarative configuration

### PR DESCRIPTION
**Pull request by jameysharp**
###### Motivation for this change

I find it difficult to understand what the dependencies are between the fragments of scripts that implement stage-2 init and activation. This branch illustrates several ways of eliminating some fragments of those scripts in favor of systemd tools like tmpfiles.d. In one case I found the snippet was not necessary at all, as far as I can tell.

As I reported in #47550, using tmpfiles.d is not entirely safe on NixOS, but so many things rely on it at this point that I think that just needs to be fixed, rather than continuing to use activation snippets to avoid the problem.

###### Things done

This series changes only modules that I'm personally using. I've tested these patches against my personal system config in QEMU, by first starting with an empty disk and verifying that first-boot comes up without errors; then rebooting the VM and verifying that stage-2-init still works with an already-initialized filesystem; and finally running `/run/current-system/bin/switch-to-configuration test` and verifying that activation on a running system works.

I don't know which NixOS tests are relevant and haven't figured out how to run any of them yet, so presumably I need to go read some documentation somewhere...

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
This is a lot of critical stuff changed here. This will need some testing.

**Comment by Mic92** [comment]
Tests can be run by calling `nix-build` on the file in questions. Some file contain test sets in which case you can run `nix-build ./file.nix -A subtest`.

**Comment by Mic92** [comment]
@GrahamcOfBorg test boot.biosCdrom boot.netboot boot.uefiUsb

**Comment by Mic92** [comment]
@GrahamcOfBorg test boot.boot-stage1

**Comment by Mic92** [comment]
@GrahamcOfBorg test boot-stage1

**Comment by Mic92** [comment]
@GrahamcOfBorg test containers-imperative

**Comment by arianvp** [comment]
I and @flokli are happy to have a closer look and review this. We were actually working on something very similar during the All Systems Go Hackfest this weekend and have some (very unstructured)  notes here. https://hackmd.io/vFdKH1PpT5CbAp45SxOlAw#

Guess the section about "Cleanup" is most relevant. Rest is just our rambling of stuff we learnt at the conference

File: nixos/modules/system/activation/activation-script.nix [hunk 1]
```diff
@@ -128,14 +128,6 @@ in
       ''
         # Various log/runtime directories.
 
-        mkdir -m 0755 -p /run/nix/current-load # for distributed builds
```
Comment (matthewbauer) [inline nixos/modules/system/activation/activation-script.nix]: This isn't done anywhere else is it?

File: nixos/modules/system/boot/systemd.nix [hunk 2]
```diff
@@ -783,19 +784,6 @@ in
 
     services.dbus.enable = true;
 
-    system.activationScripts.systemd = stringAfter [ "groups" ]
-      ''
-        mkdir -m 0755 -p /var/lib/udev
```
Comment (matthewbauer) [inline nixos/modules/system/boot/systemd.nix]: Same for this one

File: nixos/modules/system/activation/activation-script.nix (hunk 1 above)
Comment (Mic92) [reply nixos/modules/system/activation/activation-script.nix]: In the commit message it says that this was used in an very old version of nix.

File: nixos/modules/system/boot/systemd.nix (hunk 2 above)
Comment (Mic92) [reply nixos/modules/system/boot/systemd.nix]: And this was also used in pre-systemd times.

File: nixos/modules/system/activation/activation-script.nix (hunk 1 above)
Comment (matthewbauer) [reply nixos/modules/system/activation/activation-script.nix]: Ok thanks. Missed the commit messages but they are very informative.

**Comment by xaverdh** [comment]
Excellent!
So 10e865051548f39e8bff205ea09a648c49304d11 will allow actually making proper use of "Storage=auto".
Until now auto was essentially an alias of persistent ;-)

**Comment by jameysharp** [comment]
@xaverdh Yes! Assuming that commit gets merged, I think there should also be a new config option for journald.conf's Storage setting so people can pick which mode they want. I just figured this pull request was invasive enough without adding that change to it as well.

**Pull request merged**